src/Backups/BackupIO_S3.cpp

```diff
@@ -385,6 +404,58 @@ void BackupWriterS3::removeFiles(const Strings & file_names)
     }
 }
 
+void BackupWriterS3::removeFilesSingle(const Strings & file_names)
+{
+    for (const String & file_name : file_names)
+    {
+        Aws::S3::Model::DeleteObjectRequest request;
+        request.SetBucket(s3_uri.bucket);
+        request.SetKey(fs::path(s3_uri.key) / file_name);
+        auto outcome = client->DeleteObject(request);
+        if (!outcome.IsSuccess())
+            throw Exception(outcome.GetError().GetMessage(), ErrorCodes::S3_ERROR);
+    }
+}
+
+void BackupWriterS3::checkS3Access()
+{
+    const std::string_view payload("test", 4);
+    const String path = fmt::format("clickhouse_backup_objects_capability_{}", getServerUUID());
+
+    try
+    {
+        auto file = writeFile(path);
+        file->write(payload.data(), payload.size());
+        file->finalize();
+    }
+    catch (...)
+    {
+        try
+        {
+            removeFilesSingle({path});
+        }
+        catch (...)
+        {
+        }
+        throw;
+    }
+
+    try
+    {
+        removeFilesBatch({path});
+
+        LOG_TRACE(log, "Batch delete: supported");
+        supports_batch_delete = true;
+    }
+    catch (const Exception &)
+    {
+        removeFilesSingle({path});
+
+        LOG_TRACE(log, "Batch delete: unsupported");
+        supports_batch_delete = false;
+    }
```

Comment:
This is quite a difficult way to check if batch delete is supported. Maybe for `s3` disks it's ok, however it doesn't look ok here. First, if the backup was successful you have to remove only one file - and if there is one file you have no reason to use batch delete anyway (and no need to check anything too). Second, if it's necessary to remove more than one file it's still no reason to check anything in advance - just try to call removeFilesBatch() and only if it's failed - check the error code and then set `supports_batch_delete` to 0 and call `removeFilesSingle()`.

Reply:
Ok. Rebased.